test/runner/createTestServer.js

```diff
@@ -1,70 +1,186 @@
+			if ( middleware ) {
+				try {
+					middleware( req, res, next );
+				} catch ( error ) {
+					console.error( error );
+					res.writeHead( 500, { "Content-Type": "application/json" } );
+					res.end( "Internal Server Error" );
+				}
+			} else {
+				res.writeHead( 404 );
+				res.end();
+			}
+		};
+
+		next();
+	}
 
 	// Redirect home to test page
-	app.get( "/", ( _req, res ) => {
-		res.redirect( "/test/" );
+	use( ( req, res, next ) => {
+		if ( req.parsedUrl.pathname === "/" ) {
+			res.redirect( "/test/" );
+		} else {
+			next();
+		}
 	} );
 
 	// Redirect to trailing slash
-	app.use( ( req, res, next ) => {
-		if ( req.path === "/test" ) {
-			const query = req.url.slice( req.path.length );
-			res.redirect( 301, `${ req.path }/${ query }` );
+	use( ( req, res, next ) => {
+		if ( req.parsedUrl.pathname === "/test" ) {
+			res.redirect( 308, `${ req.parsedUrl.pathname }/${ req.parsedUrl.search }` );
 		} else {
 			next();
 		}
 	} );
 
 	// Add a script tag to the index.html to load the QUnit listeners
-	app.use( /\/test(?:\/index.html)?\//, ( _req, res ) => {
-		res.send(
-			indexHTML.replace(
-				"</head>",
-				"<script src=\"/test/runner/listeners.js\"></script></head>"
-			)
-		);
+	use( ( req, res, next ) => {
+		if (
+			( req.method === "GET" || req.method === "HEAD" ) &&
+			( req.parsedUrl.pathname === "/test/" || req.parsedUrl.pathname === "/test/index.html" )
+		) {
+			res.writeHead( 200, { "Content-Type": "text/html" } );
+			res.end(
+				indexHTML.replace(
+					"</head>",
+					"<script src=\"/test/runner/listeners.js\"></script></head>"
+				)
+			);
+		} else {
+			next();
+		}
 	} );
 
 	// Bind the reporter
-	app.post(
-		"/api/report",
-		bodyParser.json( { limit: "50mb" } ),
-		async( req, res ) => {
-			if ( report ) {
-				const response = await report( req.body );
-				if ( response ) {
-					res.json( response );
-					return;
-				}
+	use( async( req, res, next ) => {
+		if ( req.url !== "/api/report" || req.method !== "POST" ) {
+			return next();
+		}
+		let body;
+		try {
+			body = JSON.parse( await getRawBody( req ) );
+		} catch ( error ) {
+			if ( error.code === "ECONNABORTED" ) {
+				return;
 			}
-			res.sendStatus( 204 );
+			console.error( error );
+			res.writeHead( 400, { "Content-Type": "application/json" } );
+			res.end( JSON.stringify( { error: "Invalid JSON" } ) );
+			return;
 		}
-	);
-
-	// Handle errors from the body parser
-	app.use( bodyParserErrorHandler() );
+		const response = await report( body );
+		if ( response ) {
+			res.writeHead( 200, { "Content-Type": "application/json" } );
+			res.end( JSON.stringify( response ) );
+		} else {
+			res.writeHead( 204 );
+			res.end();
+		}
+	} );
 
 	// Hook up mock server
-	app.use( mockServer() );
+	use( mockServer() );
 
 	// Serve static files
-	app.post( "/test/data/name.html", ( _req, res ) => {
-		res.send( nameHTML );
-	} );
+	const validMimeTypes = {
+		".js": "application/javascript",
```

Comment:
Do we want to omit `.mjs` & `.cjs` on purpose? It might make sense since many HTTP servers don't serve a JS mime type for those, but I was wondering if this was deliberate. If this was, a code comment could help.

Reply:
I didn't consider serving `.mjs` and `.cjs`. I'll add a comment.